projects/hitl/software/hitl/cancontroller.py

```diff
@@ -16,200 +11,190 @@
-        
-        Designed to input messages in following format:
+        Stop a periodic message task
+        """
+        if self.can is None:
+            self.log.error("Could not stop periodic task: CAN hardware not connected")
+            return
 
-        Timestamp,arbitration_id,signals(0 to 255 value)...
+        if msg_name in self.periodic_messages:
+            self.periodic_messages[msg_name].stop()
+            del self.periodic_messages[msg_name]
+            self.log.debug(f"Stopped periodic sending of {msg_name}")
+        else:
+            self.log.error(f"Message {msg_name} is not being sent periodically")
 
+    def stop_all_periodic(self):
         """
-        # Reading and parsing csv file
-        log_file = open(path, "r")
-        csvreader = csv.reader(log_file)
-        messages = []
-        for row in csvreader:
-            # remove empty elements from list
-            row = list(filter(None, row))
-            # convert strings to integers
-            row = [int(i) for i in row]
-            # check if row is within time range
-            if row[0] >= initial_time:
-                messages.append(row)
-        log_file.close()
-        prev_time = initial_time
-
-        start_time = time.time_ns()  # initial time
-        row = 0  # initial row
-        while row < len(messages):
-            # calculate time elapsed and check if next message should be sent
-            time_elapsed = (time.time_ns() - start_time) / 1000000
-            if time_elapsed >= int(messages[row][0]):
-                # Pull out message data from csv
-                data = messages[row][2:]
-                # create CAN message
-                message = can.Message(
-                    arbitration_id=messages[row][1], data=data, is_extended_id=False
-                )
-                # send message
-                self.can_bus.send(message)
-                # increment row
-                row += 1
+        Stop all periodic message tasks
+        """
+        if self.can_bus is None:
+            self.log.error(
+                "Could not stop all periodic tasks: CAN hardware not connected"
+            )
+            return
+
+        self.can_bus.stop_all_periodic_tasks()
+        self.log.debug("Stopped all periodic tasks")
 
-    def _create_state_dictionary(self, path: str) -> None:
+    def _create_state_dictionary(self, path: str):
         """Generate self.message_of_signal and self.signals
 
-        :param str path: Path the the CAN spec file
+        Args:
+            path (str): Path to CAN spec file (DBC)
         """
         # Create database that has all the messages in the dbc file in type message
         self.db = cantools.database.load_file(path)
 
         # Iterates through messages to create state dictionaries
         message_of_signal = {}
-        signals = {}
+        signals = defaultdict(lambda: dict())
```

Comment:
Defaultdict is a type in Python that takes a callable, and then if you try to access a key in it that doesn't exist, it creates with a default value of whatever the callable returns. So in line 189, when we run `signals[msg.name][sig.name] = 0`, with a regular dict we'd get a KeyError because `signals[msg.name]` doesn't exist, but the defaultdict will see that and automatically initialize `signals[msg.name]` to an empty dict.

However, I just looked it up and turns out it doesn't actually have to be a lambda, since dict is a callable on its own, the lambda is just for when you'd need to pass some parameters in